packages/next/server/app-render.tsx

```diff
@@ -1572,7 +1583,7 @@ export async function renderToHTMLOrFlight(
   )
 
   const requestStore = {
-    headers: new ReadonlyHeaders(headersWithoutFlight(req.headers)),
+    headers: lazy(() => new ReadonlyHeaders(headersWithoutFlight(req.headers))),
```

Comment:
```js
let cachedHeaderInstance: ReadonlyHeaders | undefined;
const requestStore = {
  get headers() {
    cachedHeaderInstance ||= new ReadonlyHeaders(headersWithoutFlight(req.headers)));
	  return cachedHeaderInstance
	},
```

This could achieve both lazy and singleton, without changing the API interface.